### PR TITLE
Cap state/local deductions as fraction of AGI

### PR DESCRIPTION
**Pull request by derrickchoe**
I'm adding a reform I made a few months back for a project. I created the parameter StateLocalTax_crt, which caps state/local tax deductions as a decimal fraction of agi.

### CODE REVIEW

File: taxcalc/current_law_policy.json [hunk 1]
```diff
@@ -1124,6 +1124,28 @@
         "out_of_range_maxmsg": "",
         "out_of_range_action": "stop"
     },
+        
+    "_ID_StateLocalTax_crt": {
+        "long_name": "Maximum state and local income and sales taxes deduction as a fraction of AGI.",
```
Comment (MattHJensen) [inline taxcalc/current_law_policy.json]: Could you take a look at `_ID_Charity_crt_all` (https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/current_law_policy.json#L1262) and model your `long_name` and `description` off of that?

File: taxcalc/functions.py [hunk 2]
```diff
@@ -551,7 +554,11 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped,
                  ID_StateLocalTax_c[MARS - 1])
     c18500 = min((1. - ID_RealEstate_hc) * e18500_capped,
                  ID_RealEstate_c[MARS - 1])
-    c18300 = c18400 + c18500
+    if ID_StateLocalTax_crt < 1:
+        c18300 = min(c18400 + c18500,
+                     ID_StateLocalTax_crt * max(c00100, 0))
```
Comment (MattHJensen) [inline taxcalc/functions.py]: It looks to me like this line applies the cap to the combination of state and local income and sales taxes _and_ state, local, and foreign real estate taxes. If that is your intent, the description of the parameters should be changed.

File: taxcalc/functions.py [hunk 3]
```diff
@@ -551,7 +554,11 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped,
                  ID_StateLocalTax_c[MARS - 1])
     c18500 = min((1. - ID_RealEstate_hc) * e18500_capped,
                  ID_RealEstate_c[MARS - 1])
-    c18300 = c18400 + c18500
+    if ID_StateLocalTax_crt < 1:
```
Comment (MattHJensen) [inline taxcalc/functions.py]: I don't think you should end up needing a conditional here given that you set ID_StateLocalTax_crt at 1 under current law.

**Comment by MattHJensen** [comment]
@derrickchoe, thanks for opening a PR to add this new feature! Just commented a few lines with suggestions. 

I also recommend adding a test that confirms setting `_ID_StateLocalTax_crt` to 0 generates the same results as setting `_ID_StateLocalTax_hc` to 1. (Or explain why you think those should be different). See https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/tests/test_calculate.py#L255 for an example.

File: taxcalc/functions.py (hunk 3 above)
Comment (derrickchoe) [reply taxcalc/functions.py]: Thanks for the comments, Matt! I included the conditional to take care of rare cases where one's state and local tax deductions exceed their AGI (I think this happens when someone can write off a large business loss). I had considered setting the parameter at 9e99 under current law, but this still wouldn't take care of cases of itemizers with negative AGI. I'm more than open to suggestions which take care of the if condition.

File: taxcalc/current_law_policy.json (hunk 1 above)
Comment (derrickchoe) [reply taxcalc/current_law_policy.json]: I'll be sure to model the ID_StateLocalTax_ct description after that parameter. Thanks!

**Comment by derrickchoe** [comment]
@martinholmer 

Thanks for the suggestions, @MattHJensen. I've split the reform parameter into separate state/local and real estate tax deduction ceilings (as fractions of AGI).  I've also set the baseline value for these parameters at 9e99, to take care of cases where one's deductions exceed their AGI. I'm having a bit of trouble understanding the issues with my current code. It seems like they stem from utils.py and calculate.py, neither of which I edited. 

Is this happening because my taxcalc branch is outdated? If so, I'd appreciate any insight you might have regarding solutions. Thanks!

**Comment by martinholmer** [comment]
@derrickchoe said:
> I'm having a bit of trouble understanding the issues with my current code. It seems like they stem from `utils.py` and `calculate.py`, neither of which I edited.

The coverage test is failing because the following two lines are untested.
```
non_sum_cols = [c for c in diffs.columns if 'perc_' in c]
for col in non_sum_cols:                           <=========== UNTESTED
    diffs.loc['sums', col] = np.nan                <=========== UNTESTED
diffs.loc['sums', 'share_of_change'] = 1.0  # to avoid rounding error
```
But that's not your fault at all.  I think you can ignore that coverage test failure because these two untested lines will be removed soon as part of the resolution in Tax-Calculator of [PolicyBrain issue 73](https://github.com/OpenSourcePolicyCenter/PolicyBrain/issues/73).

@derrickchoe also asked:
> Is this happening because my taxcalc branch is outdated? 

No.  You've done a good job.  Your confusion is my fault for leaving two untested statements on the master branch.  Sorry about that.

**Comment by martinholmer** [comment]
@derrickchoe asked:
> Is this happening because my taxcalc branch is outdated? 

Again, I think the answer is no.  

But it is true that your branch does not include many recent changes on the master branch.  So, for future reference, when you have a long-running development branch on your local computer, it is good practice to incorporate recent master changes into that branch.  Do that by following the directions in **Workflow Step 6** in the [Contributor Guide](http://taxcalc.readthedocs.io/en/latest/contributor_guide.html).

Thanks for your contribution to Tax-Calculator.

**Comment by derrickchoe** [comment]
@martinholmer, thanks for clarifying. I'll be sure to keep my branches updated with recent changes to the master branch in the future. I appreciate your help!

**Comment by martinholmer** [comment]
@derrickchoe said:
> thanks for clarifying. I'll be sure to keep my branches updated with recent changes to the master branch in the future. I appreciate your help!

Sure, no problem.  There are a lot of moving parts involved in making even a modest contribution to Tax-Calculator.  You've done very well.

**Comment by martinholmer** [comment]
@MattHJensen,  Is pull request #1711 ready to be merged?  If so, go ahead and do that.

**Comment by MattHJensen** [comment]
@derrickchoe, this is excellent. Thanks. 

I do [still]((https://github.com/open-source-economics/Tax-Calculator/pull/1711#issuecomment-348690874)) recommend taking the additional step of adding a couple of tests to ensure that setting _ID_StateLocalTax_crt to 0 generates the same results as setting _ID_StateLocalTax_hc to 1 and same for real estate taxes. 

See https://github.com/open-source-economics/Tax-Calculator/blob/master/taxcalc/tests/test_calculate.py#L255 for an example.

Alternatively, you could add these (4) reforms to `reforms.json` and visually ensure that the results are the same. In general, when we have two parameters that allow us to get at the same reform in different ways, it is a great opportunity to check that everything is working as expected. 

cc @martinholmer

**Comment by derrickchoe** [comment]
@MattHJensen, I'm happy to add those tests (it will be valuable python practice as well). I'll make the changes to test_calculate.py and push them once I'm finished. Thanks!

@martinholmer

**Comment by MattHJensen** [comment]
> I'm happy to add those tests (it will be valuable python practice as well). I'll make the changes to test_calculate.py and push them once I'm finished. 

Thanks!

**Comment by derrickchoe** [comment]
@MattHJensen @martinholmer 

I've just gotten to writing in some tests to make sure that capping state/local and real estate tax deductions at 0% of AGI is equivalent to a full haircut of those respective parameters. I wrote two functions in test_calculate.py, modeled after the code you linked.

https://github.com/open-source-economics/Tax-Calculator/blob/c36faaf8b26c40af0762218f17267a049da7e4eb/taxcalc/tests/test_calculate.py#L270

I'm not too familiar with tests-- are there any other files in taxcalc that I should edit so that the tests run properly? Thanks to both of you for your help, and sorry in advance if this info was covered somewhere in the tax calc documentation.

**Comment by hdoupe** [comment]
@derrickchoe these tests look good.  You can see that they ran by running this locally:

`py.test -m "not requires_pufcsv" -v`

You can also check the output from the travis build:
screen shot 2017-12-14 at 11 18 27 am

This can be found by clicking "Show all checks" in the "All checks have passed" tab and then clicking continous-integration/travis-ci/pr.....  Finally click one of the build jobs.  Your options should be python 2.7 or python 3.6.

**Comment by derrickchoe** [comment]
@hdoupe thanks for the info--I'm not too familiar with all of the checks that run on github, so it helps to be able to isolate the individual tests I wrote. 

I just recently learned to run tests from the command line, as well.

I appreciate your help!

**Comment by hdoupe** [comment]
No problem. Good work here @derrickchoe

**Comment by martinholmer** [comment]
@derrickchoe,  We appreciate your interest in Tax-Calculator and your proposed enhancement in pull request #1711.  But this pull request --- which contains 35 changed files and 167 commits --- is basically impossible for us to review.  As far as I can tell, your proposal is to add one new policy parameter in `current_law_policy.json`, add tax logic for that parameter in `functions.py`, and add a test for that new parameter and it's logic in `test_calculate.py`.  So, your pull request should be relatively simple with 3 changed files.

I'm not absolutely sure why the current version of your pull request has so much other stuff in it, but I suspect it's related to the fact that your master branch is very out of date.  If you go to your GitHub clone of the Tax-Calculator repository, which is [here](https://github.com/derrickchoe/Tax-Calculator), you'll see this:

screen shot 2017-12-18 at 8 32 05 am

So, there have been 379 commits to the central Tax-Calculator repository since you last updated the master branch in you clone of Tax-Calculator.  Please follow the steps in the Workflow section of the [Contributor Guide](http://taxcalc.readthedocs.io/en/latest/contributor_guide.html) to get your master branch up-to-date and then merge the master branch into your `test` development branch and finally push the test branch to origin.  If the instructions in the Guide are unclear, then please ask questions here so that we can improve the Guide and get your pull request into a form that we can effectively review.

I know this is all pretty complicated.  I still remember my first weeks working on Tax-Calculator when I was new to Python and unfamiliar with git.  There were moments that were pretty confusing for me, and I still get confused sometimes (especially about git, which is powerful, but is not going to win any ease-of-use awards).  So, we hope you keep working on pull request #1711 and we look forward to your questions and feedback so we can improve our Contributor Guide.

**Comment by derrickchoe** [comment]
Hi @martinholmer,

Thanks for the update! I was unsure why my pull request involved so many changed files. I've been using the commands "fetch upstream" and "merge upstream/master" throughout this process, so it's possible that I changed something having to do with my clone of tax-calculator (I see now that it still is showing old commits that I should have discarded). @hdoupe has kindly offered his assistance in getting my workflow corrected so that these issues don't arise in the future. I'm looking forward to pushing my updated files as soon as I have everything sorted out.

**Comment by derrickchoe** [comment]
@martinholmer @MattHJensen

Thank you all for your patience-- this relatively simple addition to tax-calculator was a great introduction to git usage. As @hdoupe mentioned to me, one of the potential issues with my master branch was my failure to use "git push" to update my remote repository continually after updating my local master branch. I'm still learning how my local repository, "derrickchoe/tax-calculator.git" remote repository, and the central tax-calculator repository interact with one another. I'm looking forward to making smoother updates to tax-calculator in the future.

**Comment by martinholmer** [comment]
@derrickchoe,  Thanks for the updates to your pull request #1711.  I'll try to review it this afternoon.

**Comment by martinholmer** [comment]
I've tested pull request #1711 on my computer and there are failures in two tests that use the `puf.csv` input file, which @derrickchoe does not have available to him.  The first is in the `pufcsv_agg_expect.txt` output for `AMI Income` in 2017 (was 10,896.6 and now is 10,896.5) and 2021 (was 12,613.9 and now is 12,613.8).  And the second is in the `puf_var_wght_means_by_year.csv` file for `c04470` in several years (with the changes being quite small).  It seems as if these slight differences are caused by the insertion in #1711 of these two lines in the `ItemDed` function:
```
    c18400 = min(c18400, 
                 ID_StateLocalTax_crt * max(c00100, 0))
    c18500 = min(c18500,
                 ID_RealEstate_crt * max(c00100, 0))
```
It seems to me these slight differences are rounding errors (rather than logic errors), and if I don't hear any objections, I'm going to merge #1711 at the end of the work day on Monday, Dec 18th.

@MattHJensen @Amy-Xu @hdoupe @andersonfrailey @GoFroggyRun @codykallen

**Comment by codykallen** [comment]
@martinholmer said

> It seems to me these slight differences are rounding errors (rather than logic errors)

I'm not sure that's necessarily true. If a filing unit has negative AGI (e.g. due to writing off a large business loss), but this filing unit has some positive `e18500`, then this individual would have `c18500 == 0`, whereas under current law he would have `c18500 > 0`. 

Normally, an individual with negative AGI would not have taxable income anyway, so this would not matter for calculating taxable income. However, it would matter for calculating AMT taxable income, because the AMT calculations have different treatments of net losses, and the AMT adds back the state and local income/sales/property taxes. 

Therefore, for individuals with negative AGI but positive `e18400` or `e18500`, this would reduce the total amount of `c18400` and `c18500` and thus the total amount of `c04470`. Because `c18400` and `c18500` are added back in the AMT calculations, this also reduces AMT taxable income.

**Comment by derrickchoe** [comment]
@martinholmer 
I should mention that I do have the puf.csv file (Anderson gave me access when I started working at AEI). @codykallen makes a good point-- I forgot that people with negative AGIs can still have tax liabilities. To ensure that everything functions as it should under current law, I can change my code to minimize 

` c18400 = min(c18400, 
                 ID_StateLocalTax_crt * max(c00100, 0))`

**Comment by codykallen** [comment]
@martinholmer, what do you think about changing the code to:
`c18400 = min(c18400, ID_StateLocalTax_crt * max(c00100, 0.0001))`

I don't love this solution, but I think it would be an effective means to avoid the issue of those with negative AGI.

**Comment by GoFroggyRun** [comment]
@codykallen, I was testing the solution you proposed, and it doesn't appear to cause any test failure:
  
```
c18400 = min(c18400, ID_StateLocalTax_crt * max(c00100, 0.0001))
c18500 = min(c18500, ID_RealEstate_crt * max(c00100, 0.0001))
```

**Comment by martinholmer** [comment]
@codykallen,  Thanks for your astute observation, which I agree with (now that you've pointed out what's going on in the logic).  The small changes in results are just as would be expected given your explanation.

So, the only remaining question is how to fix pull request #1711.

@derrickchoe suggested:
> To ensure that everything functions as it should under current law, I can change my code to only minimize state/local and real estate tax deductions if the [two] `_crt` parameters are specified [to be] smaller than 1.

I guess by this, he means code like this:
```
if ID_StateLocalTax_crt <= 1.0:
    c18400 = min(c18400, ID_StateLocalTax_crt * posagi)
if ID_RealEstate_crt <= 1.0:
    c18500 = min(c18500, ID_RealEstate_crt * posagi)
```
@codykallen suggested getting (I think) the same result without the if statements by having the `_crt` parameters multiplied by `max(c00100, 0.0001)` instead of by `posagi`.  But he said "I don't love this solution", presumably because its not immediately obvious why the 0.0001 is being used.

So, is this a correct understanding of what everybody's saying here?  I'm I right or wrong that these two solutions are mathematically equivalent?  What should be do?

Thanks again for all the quick, thoughtful comments.  Looking forward to the next round of comments.

**Comment by martinholmer** [comment]
@derrickchoe,  If you have the `puf.csv` file, then by all means run the full set of tests on your computer.

**Comment by derrickchoe** [comment]
@martinholmer that's exactly what I mean. I do slightly prefer @codykallen 's solution. If nobody objects, I'm happy to change the code push the new changes. 

I'll be sure to run the tests on my computer in the future. I was eager to confirm that my updated test branch was no longer pushing 160+ commits, and I neglected to run the tests. For future reference, will `py.test -m "not requires_pufcsv" ` be sufficient to catch the errors outlined above? I only ask because I know the full tests take significantly longer to complete.

**Comment by martinholmer** [comment]
@derrickchoe asked:
> For future reference, will `py.test -m "not requires_pufcsv"` be sufficient to catch the errors outlined above? I only ask because I know the full tests take significantly longer to complete.

No, that won't be sufficient.  The two test failures I got were for `requires_pufcsv` tests.  
The full tests do take somewhat longer, but you can get much faster execution by using the parallel processing feature of `py.test` to execute the full tests this way:
```
py.test -n4
```

**Comment by martinholmer** [comment]
@derrickchoe said:
>  I do slightly prefer @codykallen 's solution. If nobody objects, I'm happy to change the code [and] push the new changes.

Fine, why don't you do that and see if your changes pass **all** the tests before you push.   Thanks.

**Comment by derrickchoe** [comment]
@MattHJensen 

I'm having trouble getting the full tests to work when swapping 0.0001 for 0 in the following code:
```
    c18400 = min(c18400, 
                 ID_StateLocalTax_crt * max(c00100, 0.0001))
    c18500 = min(c18500,
                 ID_RealEstate_crt * max(c00100, 0.0001))
```
The issues are coming with the test_pufcsv.py file. I'm going to try using conditional statements instead (as @martinholmer outlined above). Thank you all for your patience; I'm hoping that I can have a working pull request tomorrow.

**Comment by martinholmer** [comment]
@derrickchoe said:
> I'm having trouble getting the full tests to work when swapping 0.0001 for 0 in the following code:
> ...
> The issues are coming with the `test_pufcsv.py` file.

Can you show us exactly what the differences are between the actual and expected output?

**Comment by martinholmer** [comment]
@derrickchoe said:
> I'm having trouble getting the full tests to work when swapping 0.0001 for 0 in the following code:
> ...
> The issues are coming with the `test_pufcsv.py` file.

Are you running Python 3.6 (rather than 2.7)?

**Comment by martinholmer** [comment]
@derrickchoe, I just downloaded PR#1711 and changed the two `max(c00100, 0)` expressions to `max(c00100, 0.0001)` per the suggestion made by @codykallen.  All the tests (including the `requires_pufcsv` tests) run without any failures.

Can you make those coding changes (like @codykallen suggested) and commit them and push them, so that I can merge #1711?   We'll figure out later what sort of problems you're having with the tests.  But both @GoFroggyRun and I have confirmed that the @codykallen revisions eliminate all the test failures that I reported yesterday.

One additional question:  when you are not on GitHub would you say your name is Derrick Choe?
If not, what you you say your name is?

**Comment by derrickchoe** [comment]
@martinholmer 

Thanks for asking-- yes my name is Derrick Choe. 

In response to: 
> Are you running Python 3.6 (rather than 2.7)?

I believe I am running Python 3.6. I did install a Python 2.7 environment via Anaconda, though. I'm still familiarizing myself with this, and I'll be sure to confirm exactly with what version of Python I'm running tax-calculator.

In response to:

> Can you show us exactly what the differences are between the actual and expected output?

I'll pull up the test error message shortly. I also had a couple of issues with index.html (which I know I can fix by running make_index.py), as well as an error message related to a missing mock module, but I am told that these are not problems with the code itself.

Lastly:

> Can you make those coding changes (like @codykallen suggested) and commit them and push them, so that I can merge #1711?

I'll be sure to do that now. Thanks again for all of your guidance with this pull request!

**Comment by martinholmer** [comment]
Thanks, @derrickchoe, for the enhancements in pull request #1711.

**Comment by derrickchoe** [comment]
@martinholmer 
I just reran the tests using my Python 2.7 environment console. The errors associated with test_pufcsv.py did not show up this time. Sorry about that-- I'll be sure to use Python 2.7 for future tests.

**Comment by martinholmer** [comment]
@derrickchoe said:
> I just reran the tests using my Python 2.7 environment console. The errors associated with `test_pufcsv.py` did not show up this time.   Sorry about that-- I'll be sure to use Python 2.7 for future tests.

I think your experience using Python 3.6 is the same as @MattHJensen's experience in #1715.  There seems to be slight differences in the floating-point libraries in 2.7 vs 3.6, which produce slightly different numerical results.  Can you post the `test_pufcsv.py` differences you get when running Python 3.6?  Are they the same as @MattHJensen reports in #1715?

So, you didn't do anything wrong and there's nothing to be sorry for.  Thanks again for this contribution.

**Comment by derrickchoe** [comment]
@martinholmer  I get the following results when I run the test using Python 3.6 (please let me know if the screenshots are too messy-- I'm not exactly sure how to best show failed tests):

python3 6_error


python3 6_error1

Is this the output that you are referring to? Looking at #1715, I am having a bit of trouble identifying exactly which differences are arising in the tests.

**Comment by martinholmer** [comment]
@derrickchoe,  Thanks for the info on test failures under Python 3.6.  
Looks like you're working on Windows, right?

When you ran these tests your master branch was up-to-date, right?
Did it include the changes in pull request #1773?  If not, please update your master branch so that it includes the 1773 changes.

What's the difference between the `pufcsv_agg_expect.txt` file and the `pufcsv_agg_actual.txt` file?
Get the differences in the two files using a diff utility.   Working on Windows means you don't have many useful tools available at the command prompt, so you might need to find a Windows diff utility and install it on your computer.

**Comment by derrickchoe** [comment]
@martinholmer 

While I had forgotten to update my master branch with changes in pull request #1773, I am getting the same test failures with my now-updated branch. The files `test_pufcsv.py` and `test_puf_var_stats.py` are both showing the same errors. Additionally, I'm getting an issue with `test_reforms.py`-- I think the issue has to do with the rename function (output below):


python3 6_error2

In response to:

> What's the difference between the pufcsv_agg_expect.txt file and the pufcsv_agg_actual.txt file?
Get the differences in the two files using a diff utility.

I'm not too familiar with diff utilities; I'm looking into installing and using one now. I'll send an update once I have a better understanding of the differences between the two files. Thanks!

**Comment by martinholmer** [comment]
Thanks for the report on the test failure in `test_reforms.py`.
Resolve that part of the problem by reading pull request #1781.

**Comment by derrickchoe** [comment]
@martinholmer @MattHJensen 
image

Sorry in advance for the large screenshot. I'm getting differences between actual and expected output for the following numbers: 

2018 AMT liability
2019 itemized deductions and taxable income
2020 taxable income

All of these differences are in the magnitude of 0.1 billion. Is it possible that this is a rounding error?

**Comment by martinholmer** [comment]
@derrickchoe said:
> I'm getting differences between actual and expected output for the following numbers:
>
> 2018 AMT liability
> 2019 itemized deductions and taxable income
> 2020 taxable income
>
> All of these differences are in the magnitude of 0.1 billion. Is it possible that this is a rounding error?

Yes, I think it is just rounding error caused by differences in the floating-point libraries between Python 2.7 and Python 3.6.   We have experienced this before with other tests that use the `cps.csv` input, which are run under both 2.7 and 3.6 when you create a GitHub pull request.

These test failures are happening even on the master branch, right?

I'll try to generalize `test_pufcsv.py` so it doesn't fail when actual and expected differ by only $0.1 billion.
When I do that, I'll ask you to run the tests on your computer to see if they all pass.  (Remember GitHub doesn't run the `requires_pufcsv` tests because of data privacy issues.)

Thanks for all the help!

**Comment by derrickchoe** [comment]
@martinholmer 
In response to:

> These test failures are happening even on the master branch, right?

That's right; I ran these tests on my up-to-date master branch. Thanks for addressing this; I'll be sure to stay in the loop. I'm happy to run the tests whenever you think is appropriate. In the meantime, I'm glad that the recent additions to tax-calculator are running smoothly! I'm looking forward to making future contributions.

**Comment by martinholmer** [comment]
@derrickchoe,  Can you update your master branch to include several recent merges on GitHub, and then run all the tests including the `requires_pufcsv` tests.  Then let me know if all the 390+ tests pass on your computer.  If one or more failed, let me know what the differences are.

Thanks so much in advance for helping me out on this.  I don't have Python 3.6, so I'm thankful that you do and that you are willing to run the tests.

**Comment by derrickchoe** [comment]
@martinholmer 

I've updated my master branch, and `test_pufcsv.py` is working now. However, I'm still getting an error message with `test_puf_var_stats.py`. I've attached the differences below (they occur in lines 5-6 of `puf_var_wght_means_by_year.csv`).

image

I'm happy to help out-- given enough time working with tax-calculator, I hope to be able to help fix minor issues (such as these ones) in the future as well.

**Comment by derrickchoe** [comment]
@martinholmer 
In response to:

> If one or more failed, let me know what the differences are.

I should mention that I'm working on my personal computer now, and I noticed that a new test failed (I suspect it just has to do with my installation of taxcalc).

In `test_4package.py`, I'm getting a "file not found" error for `lib\subprocess.py` (I don't know if this means that subprocess.py is missing). Is this a common issue?

**Comment by martinholmer** [comment]
@derrickchoe said:
> In test_4package.py, I'm getting a "file not found" error for lib\subprocess.py (I don't know if this means that subprocess.py is missing). Is this a common issue?

No its not common, but more on that below.

My understanding is that subprocess is part of the standard, built-in Python library.
So, if you fail the tests in `test_4package.py`, then you must also be failing the tests in `test_notebooks.py`.  Is that the case?  I'm guessing that because the subprocess library is imported in both those files as seen below:
```
- grep; default-directory: "~/work/OSPC/tax-calculator/taxcalc/tests/" -*-
grep -nH -e subprocess *
test_4package.py:10:import subprocess
test_4package.py:22:    out = subprocess.check_output(['conda', 'list', 'taxcalc']).decode('ascii')
test_notebooks.py:3:import subprocess
test_notebooks.py:16:    rcode = subprocess.check_call(['python', sfilename], cwd=test_path)
```
Nobody has ever reported this as a problem, but maybe that's because all the other developers are working on Macs and you're working on Windows.  That's just a guess.

Are you executing `py.test -n4` from the Windows command prompt or from inside some IDE?

**Comment by codykallen** [comment]
@martinholmer said:

> Nobody has ever reported this as a problem, but maybe that's because all the other developers are working on Macs and you're working on Windows. 

I don't think that's the issue. I'm working on Windows, and I've never had this problem.

**Comment by martinholmer** [comment]
@derrickchoe said:
> I've updated my master branch, and `test_pufcsv.py` is working now. However, I'm still getting an error message with `test_puf_var_stats.py`. I've attached the differences below (they occur in lines 5-6 of `puf_var_wght_means_by_year.csv`).

Thanks for the helpful feedback.  I'll fix these problems in my next pull request.  And then I'm hoping you can run the tests again to see that all the test failures caused by numerical results differences have been eliminated.

**Comment by martinholmer** [comment]
@codykallen said:
> I don't think that's the issue. I'm working on Windows, and I've never had this problem.

OK, I didn't know you were working on Windows, @codykallen.  Are you using Python 2.7 or Python 3.6?

So @derrickchoe's problem seems related to the Python 3.6 setup on his personal computer.

**Comment by codykallen** [comment]
@martinholmer, I use Python 2.7 on my work computer and Python 3.6 on my personal computer.

**Comment by martinholmer** [comment]
@derrickchoe,  Hopefully with the merge of PR #1789 there will be no small numerical differences when you run the full set of tests under Python 3.6.  Thanks again for helping me out on this.

**Comment by derrickchoe** [comment]
@martinholmer in response to:

> Are you executing py.test -n4 from the Windows command prompt or from inside some IDE?

I'm executing py.test -n4 from the command line. Specifically, I'm using the taxcalc-dev environment in Anaconda Prompt. For some reason, the -n4 specification only works when working from this environment.

> So @derrickchoe's problem seems related to the Python 3.6 setup on his personal computer.

Thanks @codykallen for comparing with your installation-- I'll continue looking into this issue. I'm only failing the `test_4package.py` test.

**Comment by derrickchoe** [comment]
@martinholmer 

In response to:

> @derrickchoe, Hopefully with the merge of PR #1789 there will be no small numerical differences when you run the full set of tests under Python 3.6. Thanks again for helping me out on this.

I'm no longer running into any issues besides the one involving subprocess.py. Thanks for resolving all of the issues that I was running into using python 3.6!

**Pull request merged**